adapters/generate.py

```diff
@@ -51,25 +48,37 @@
 
     run = """
     function _{name}_run {{
-        mkdir -p out
+        mkdir -p out/{name}
     """.format(name=name)
 
     cnt = 1
     for case in data[name]['benchmarks']['cases']:
+        tar = "$W/_{name}_case{cnt}.tar.gz".format(name=name, cnt=cnt)
         prepare += """
             mkdir -p $W/_{name}_case{cnt}
-            curl {url} > $W/_{name}_case{cnt}.tar.gz
-            tar -xvzf $W/_{name}_case{cnt}.tar.gz -C $W/_{name}_case{cnt}
-        """.format(url=case['data'], name=name, cnt=cnt)
+            if [ -f {tar} ]; then
+                hashed=$(md5sum {tar} | grep -o '^[^ ]*')
+                hdr=$(curl -I {url} | grep -i Etag | grep -o ' ".*"' | grep -o '[^"]*')
+                if [ $hdr == $hashed ]; then
+                    echo "Data `{tar}` already exists with correct md5sum. It is not necessary to download with curl again."
```

Comment:
We need to extract the data from the archive, because benchmark tests can modify some data in target directory and we need always fresh copy of data. So the line `tar -xzf ...` can be moved after `if` condition